feat(core): Ability to disable minimize/maximize/close native window's buttons, closes #2353

TDiblik opened the pull request:
### What kind of change does this PR introduce?
- [ ] Bugfix
- [x] Feature ---- This PR resolves issue/feature request #2353 
- [ ] Docs
- [ ] New Binding issue #___
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

### Does this PR introduce a breaking change?
- [ ] Yes, and the changes were approved in issue #___
- [x] No

### Checklist
- [x] When resolving issues, they are referenced in the PR's title (e.g `fix: remove a typo, closes #___, #___`)
- [x] A change file is added if any packages will require a version bump due to this PR per [the instructions in the readme](https://github.com/tauri-apps/tauri/blob/dev/.changes/readme.md).
- [x] I have added a convincing reason for adding this feature, if necessary ---- This PR resolves issue/feature request #2353 

### Other information
This PR intentionally exposes functionality only to rust api (from tao). I believe that since this functionality manipulates native window's titlebar buttons state, it should not be accessible from TypeScript (although I'll be more than happy to implement it if necessary/you wish)

TDiblik committed:
feat(core): expose maximizability/minimizability/closability to window builder and attributes

TDiblik committed:
feat(utils): ability to set button states from tauri.conf.json

TDiblik committed:
chore: summarize changes

TDiblik committed:
chore: docs

TDiblik committed:
chore: add generated schemas

TDiblik committed:
chore: docs

amrbashir commented on core/tauri/src/window.rs:
```diff
@@ -1016,6 +1039,21 @@ impl<R: Runtime> Window<R> {
     self.window.dispatcher.is_resizable().map_err(Into::into)
   }
 
+  /// Gets the window’s native maximize button state (enabled/disabled)
+  pub fn is_maximizable(&self) -> crate::Result<bool> {
+    self.window.dispatcher.is_maximizable().map_err(Into::into)
+  }
+
+  /// Gets the window’s native minimize button state (enabled/disabled)
+  pub fn is_minimizable(&self) -> crate::Result<bool> {
+    self.window.dispatcher.is_minimizable().map_err(Into::into)
+  }
+
+  /// Gets the window’s native close button state (enabled/disabled)
+  pub fn is_closable(&self) -> crate::Result<bool> {
+    self.window.dispatcher.is_closable().map_err(Into::into)
+  }
+
```
For the sake of consistency and I feel this is already infered from the method name and its return type.
```suggestion
  /// Gets the window’s native maximize button state
  pub fn is_maximizable(&self) -> crate::Result<bool> {
    self.window.dispatcher.is_maximizable().map_err(Into::into)
  }

  /// Gets the window’s native minimize button state
  pub fn is_minimizable(&self) -> crate::Result<bool> {
    self.window.dispatcher.is_minimizable().map_err(Into::into)
  }

  /// Gets the window’s native close button state
  pub fn is_closable(&self) -> crate::Result<bool> {
    self.window.dispatcher.is_closable().map_err(Into::into)
  }

```

amrbashir reviewed (changes requested):
Nice work, thank you for the contribution.

A couple of things to note:

1. It is fine to expose these methods in typescript as well as the window config options.
2. You seem to have left the platform-specific documentation from tao, We need to document them in tauri as well.

TDiblik committed:
chore(core): make comments consistent

Co-authored-by: Amr Bashir <amr.bashir2015@gmail.com>

TDiblik commented:
> Nice work, thank you for the contribution.
> 
> A couple of things to note:
> 
> 1. It is fine to expose these methods in typescript as well as the window config options.
> 2. You seem to have left the platform-specific documentation from tao, We need to document them in tauri as well.

Hi, thanks for the feedback!

1. Will do! --- UPDATE: (Done, requesting re-review, since I'm unsure about the second point)
2. Could you please expand on this point? I made the docs by "copying/replicating" the style of `resizable` methods

TDiblik committed:
feat(core/api): Expose new api for typescript and config features

TDiblik committed:
chore: docs

TDiblik committed:
chore(core/api): add generated schemas

TDiblik committed:
chore(meta): update changes to reflect current state

TDiblik committed:
chore(core/api): add generated files

This commit was made from my notebook, instead of my pc, because windows was generating these files weirdly

amrbashir commented:
The tao functions you're trying to expose contain platform-specific information in its documentation like this one https://docs.rs/tao/latest/tao/window/struct.Window.html#method.set_closable and we need to document the same here as well.

TDiblik committed:
fix(ci)

TDiblik committed:
chore(docs): Add platform specific documentation

TDiblik committed:
chore(tooling): Rename/Remove params inside docs that don't match function

This commit is outside of the PR focus, however I saw some docs that did not match the function, so I fixed it

TDiblik committed:
chore(core/api) add generated files

TDiblik commented:
Hi, thanks for more info!

I've updated docs based on your previous comment (added platform specific information). Docs inside tauri and tao should now match.

amrbashir commented on core/tauri-runtime/src/webview.rs:
```diff
@@ -140,9 +140,35 @@ pub trait WindowBuilder: WindowBuilderBase {
+  /// Whether the window's native maximize button is enabled or not.
+  /// If resizable is set to false, this setting is ignored.
+  ///
+  /// See [`Dispatch::set_maximizable`] for details.
+  ///
+  /// [`Dispatch::set_maximizable`]: crate::Dispatch::set_maximizable
```
I'd rather embed the documentation here instead of linking to another place.

TDiblik replied:
[Fixed](https://github.com/tauri-apps/tauri/pull/6665/commits/51697f033c67ca5e6ef0f105ca83d880228d7e10)

amrbashir commented on core/tauri-utils/src/config.rs:
```diff
@@ -919,9 +919,18 @@ pub struct WindowConfig {
+  /// Whether the window is resizable or not. When resizable is set to false, native window's maximize button is automatically disabled.
   #[serde(default = "default_resizable")]
   pub resizable: bool,
+  /// Whether the window's native maximize button is enabled or not. If resizable is set to false, this setting is ignored. Note that this setting is not supported on Linux, iOS and Android.
```
you also need to add the platform-specific notes here as well

TDiblik replied:
Should the docs in this file have the same style as docs in the rest of the PR? I wrote them out differently, because I saw that people wrote platform-specific comments at the end of the line as a "note" in this file.

Examples: [1](https://github.com/tauri-apps/tauri/blob/dfb5f52d3cbe44ce4607cd875b9632093b9123b1/core/tauri-utils/src/config.rs#L955) ;  [2](https://github.com/tauri-apps/tauri/blob/dfb5f52d3cbe44ce4607cd875b9632093b9123b1/core/tauri-utils/src/config.rs#L958) ; [3](https://github.com/tauri-apps/tauri/blob/dfb5f52d3cbe44ce4607cd875b9632093b9123b1/core/tauri-utils/src/config.rs#L963)

Do you recon I rewrite it in the style [A](https://github.com/tauri-apps/tauri/blob/dfb5f52d3cbe44ce4607cd875b9632093b9123b1/core/tauri-utils/src/config.rs#L936) or [B](https://github.com/tauri-apps/tauri/pull/6665/files#diff-a188a5964932c9b690a134501222d34882286e8d69a3f70534d09108822ddab4R666)?

amrbashir replied:
lets use this style https://github.com/tauri-apps/tauri/pull/6665/files#diff-a188a5964932c9b690a134501222d34882286e8d69a3f70534d09108822ddab4R681-R685 since it there is more info that need to be explained other than just what platform is supported.

TDiblik replied:
[Fixed](https://github.com/tauri-apps/tauri/pull/6665/commits/e4e6bda03dd6444bf118641dd41c23e4f549638d)

amrbashir commented on core/tauri/src/window.rs:
```diff
@@ -393,12 +393,47 @@ impl<'a, R: Runtime> WindowBuilder<'a, R> {
+  /// Whether the window's native maximize button is enabled or not.
+  /// If resizable is set to false, this setting is ignored.
+  ///
+  /// See [`Window::set_maximizable`] for details.
+  ///
+  /// [`Window::set_maximizable`]: crate::window::Window::set_maximizable
```
same, embed the docs here

TDiblik replied:
[Fixed](https://github.com/tauri-apps/tauri/pull/6665/commits/480176869123760ba1d9ff4543b6629fb9942762)

amrbashir commented on tooling/api/src/window.ts:
```diff
@@ -683,6 +686,93 @@ class WindowManager extends WebviewWindowHandle {
+   *
+   * #### Platform-specific
+   *
+   * - **Linux / iOS / Android:** Unsupported.
+   * @example
```
```suggestion
   * - **Linux / iOS / Android:** Unsupported.
   *
   * @example
```

amrbashir commented on tooling/api/src/window.ts:
```diff
@@ -683,6 +686,93 @@ class WindowManager extends WebviewWindowHandle {
+   * #### Platform-specific
+   *
+   * - **Linux / iOS / Android:** Unsupported.
+   * @example
```
```suggestion
   * - **Linux / iOS / Android:** Unsupported.
   *
   * @example
```

amrbashir commented on tooling/api/src/window.ts:
```diff
@@ -683,6 +686,93 @@ class WindowManager extends WebviewWindowHandle {
+   * #### Platform-specific
+   *
+   * - **Linux / iOS / Android:** Unsupported.
+   * @example
```
```suggestion
   * - **Linux / iOS / Android:** Unsupported.
   *
   * @example
```

amrbashir commented on tooling/api/src/window.ts:
```diff
@@ -865,6 +954,102 @@ class WindowManager extends WebviewWindowHandle {
+   * #### Platform-specific
+   *
+   * - **macOS:** Disables the "zoom" button in the window titlebar, which is also used to enter fullscreen mode.
+   * - **Linux / iOS / Android:** Unsupported.
+   * @example
```
```suggestion
   * - **Linux / iOS / Android:** Unsupported.
   *
   * @example
```

amrbashir commented on tooling/api/src/window.ts:
```diff
@@ -865,6 +954,102 @@ class WindowManager extends WebviewWindowHandle {
+   * #### Platform-specific
+   *
+   * - **Linux / iOS / Android:** Unsupported.
+   * @example
```
```suggestion
   * - **Linux / iOS / Android:** Unsupported.
   *
   * @example
```

amrbashir commented on tooling/api/src/window.ts:
```diff
@@ -865,6 +954,102 @@ class WindowManager extends WebviewWindowHandle {
+   * #### Platform-specific
+   *
+   * - **Linux:** GTK+ will do its best to convince the window manager not to show a close button. Depending on the system, this function may not have any effect when called on a window that is already visible
+   * - **iOS / Android:** Unsupported.
+   * @example
```
```suggestion
   * - **iOS / Android:** Unsupported.
   *
   * @example
```

TDiblik committed:
chore(api): formatting

Co-authored-by: Amr Bashir <amr.bashir2015@gmail.com>

TDiblik committed:
chore(api): formatting

Co-authored-by: Amr Bashir <amr.bashir2015@gmail.com>

TDiblik committed:
chore(api): formatting

Co-authored-by: Amr Bashir <amr.bashir2015@gmail.com>

TDiblik committed:
chore(api): formatting

Co-authored-by: Amr Bashir <amr.bashir2015@gmail.com>

TDiblik committed:
chore(api): formatting

Co-authored-by: Amr Bashir <amr.bashir2015@gmail.com>

TDiblik committed:
chore(api): formatting

Co-authored-by: Amr Bashir <amr.bashir2015@gmail.com>

TDiblik committed:
docs(core): Insert platform specific comments instead of linking

TDiblik committed:
docs(core): Insert platform specific comments instead of linking

TDiblik committed:
chore(tooling): Add more info to configuration docs

amrbashir reviewed (dismissed):
Thank you very much

TDiblik commented:
NP! Thank you for reviewing, referencing materials and overall smooth PR flow!

lucasfernog committed:
Merge remote-tracking branch 'origin/dev' into feature/2353-set-titlebar-buttons-state

lucasfernog committed:
split change files

lucasfernog committed:
update docs

lucasfernog committed:
fix covector

lucasfernog committed:
update api example
